### PR TITLE
Implement video player Binding and ACL functionality

### PR DESCRIPTION
**Pull request by chrisdecenzo**
#### Problem
1. BLE commissioning of tv-app does not work when commissioner is enabled
2. ContentLaunch from tv-casting-app broken due to new ACL enforcement

### Change overview
1. Add management of transport to BLE code
2. Add ACLs for casting app

#### Testing
How was this tested? (at least one bullet point required)
1. Verify BLE commissioning of tv-app works
2. Verify tv-casting-app call to contentLauncher succeeds

### CODE REVIEW

File: examples/tv-app/linux/main.cpp [hunk 1]
```diff
@@ -114,24 +116,71 @@ class MyPincodeService : public PincodeService
 };
 MyPincodeService gMyPincodeService;
 
+// extern CHIP_ERROR LogAccessControlEvent(const Access::AccessControl::Entry & entry,
```
Comment (andy31415) [inline examples/tv-app/linux/main.cpp:119]: delete commented out code

File: examples/tv-app/linux/main.cpp [hunk 2]
```diff
@@ -114,24 +116,71 @@ class MyPincodeService : public PincodeService
...
+        FabricIndex f;
+        Access::Privilege p;
+        entry.GetAuthMode(a);
+        entry.GetFabricIndex(f);
+        entry.GetPrivilege(p);
+        ChipLogProgress(Controller, "Entry authmode=%d fabricIndex=%d privilege=%d", static_cast<int>(a), f, static_cast<int>(p));
+
+        GetAccessControl().CreateEntry(nullptr, entry, &fabricIndex);
+
+        // SubjectDescriptor subjectDescriptor = { .fabricIndex = fabricIndex, .authMode = AuthMode::kCase, .subject = nodeId };
```
Comment (andy31415) [inline examples/tv-app/linux/main.cpp:175]: delete commented out code or add some description about why it is there (e.g. some todo explaining what the template code is supposed to be used for or when uncommenting is apropriate)

File: src/transport/raw/BLE.h [hunk 3]
```diff
@@ -91,6 +100,14 @@ class DLL_EXPORT BLEBase : public Base, public Ble::BleLayerDelegate
 
     CHIP_ERROR SetEndPoint(Ble::BLEEndPoint * endPoint) override;
 
+    /**
+     * Change BLE transport to this
+     *
+     * This is relevant when a device is both a commissioner and a commissionee
+     * and so BLE transport needs to be able to toggle between these
+     */
+    void OverrideTransport() { mBleLayer->mBleTransport = this; }
```
Comment (andy31415) [inline src/transport/raw/BLE.h:109]: Without the doxycomment, the code of `foo->OverrideTransport` is unclear. Could we name it something that does not require doccomment reads on 'what transport? override with what?'.

I suggest `SetBleLayerTransportToSelf` or provide a `& GetBleLayerTransport()` method and then code can read `foo->GetBleLayerTransport() = foo` or some other way (like `auto layer = foo->GetBleLayer(); layer->mBleTransport = foo`).

The lifetime of this override is unclear to me because it is circular:
  - mBleLayer is a member of BLEBase
  - mBLELayer will have a member pointing to BLEBase

**Review by andy31415 — DISMISSED** [review]
Please use some form of RAII to scope ble layer overrides or make the lifetime of BLE layers clear. As it stands I see:

BLEBase contains a BLELayer
BLELayer would have a member of BLE base.

having this kind of circular reference will make code very hard to maintain.

File: src/transport/raw/BLE.h (hunk 3 above)
Comment (andy31415) [reply src/transport/raw/BLE.h:109]: Reading the delta now, I see that this circular dependency is not new. 

However reading more: where is this `OverrideTransport` method called?

File: src/transport/raw/BLE.h [hunk 4]
```diff
@@ -48,8 +48,17 @@ class BleListenParameters
...
+    BleListenParameters & SetOverrideExistingTransport(bool overrideExistingTransport)
+    {
+        mOverrideExistingTransport = overrideExistingTransport;
+
+        return *this;
+    }
+
 private:
     Ble::BleLayer * mLayer;
+    bool mOverrideExistingTransport = false;
```
Comment (andy31415) [inline src/transport/raw/BLE.h:61]: This changes previous behaviour. Is the previous behaviour wrong across the board? I think BLE used to work.

File: src/transport/raw/BLE.cpp [hunk 5]
```diff
@@ -66,8 +66,16 @@ CHIP_ERROR BLEBase::Init(const BleListenParameters & param)
     VerifyOrReturnError(mState == State::kNotReady, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnError(bleLayer != nullptr, CHIP_ERROR_INCORRECT_STATE);
 
-    mBleLayer                           = bleLayer;
-    mBleLayer->mBleTransport            = this;
+    mBleLayer = bleLayer;
+    if (mBleLayer->mBleTransport == nullptr || param.HasOverrideExistingTransport())
+    {
+        mBleLayer->mBleTransport = this;
```
Comment (andy31415) [inline src/transport/raw/BLE.cpp:72]: please also update ClearState so that it clears mBleLayer->mBleTransport. I think leaving the dangling pointer there is not correct.

File: src/transport/raw/BLE.h [hunk 6]
```diff
@@ -48,8 +48,17 @@ class BleListenParameters
 
     Ble::BleLayer * GetBleLayer() const { return mLayer; }
 
+    bool HasOverrideExistingTransport() const { return mOverrideExistingTransport; }
```
Comment (andy31415) [inline src/transport/raw/BLE.h:51]: Please find some naming that makes it clearer from reading code on what we are overriding and with what.

Maybe invert and call `PreserveExistingBleLayerTransport`.

Add doc-comments on when one would want to override or preserve the BLE layer transport. Is  this class intended to work at all when ble layer transport is not overriden? If we do not override the transport, I would expect any received BLE packets to not be received, so the transport does not fully work.

**Review by andy31415 — DISMISSED** [review]
Please split into 2 PRs: one for ACL and one for BLE.

ACL is reasonably clear - add a working ACL and some todos.
BLE is unclear to me: what is the problem, how is this fix fixing it, how is a BLE transport supposed to work if it does not override receive? is it for send only (and if so maybe the methods should be 'send-only' or be some flag of send or send/receive or R/RW) or something else?

File: src/transport/raw/BLE.h (hunk 6 above)
Comment (andy31415) [reply src/transport/raw/BLE.h:51]: In hindsight, depending on sideffect, I believe not overriding make the transport write-only (since no packet received callbacks). Is that true? documentation is a must.

**Comment by Byungjoo-Lee** [comment]
LG verified BLE working well.

**Review by Byungjoo-Lee — APPROVED** [review]
LG verified BLE working well.

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 7]
```diff
@@ -388,14 +389,131 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+
+    for (const auto & binding : bindings)
+    {
+        Binding::Structs::TargetStruct::Type entries[1];
+        entries[0] = binding;
+        BindingListTypeInfo bindingList(entries);
+
+        ChipLogProgress(Controller, "Attempting to create Binding");
+        ReturnErrorOnFailure(
+            CreateBindingWithCallback(targetDeviceProxy, kTargetBindingClusterEndpointId, bindingList, successCb, failureCb));
```
Comment (bzbarsky-apple) [inline src/app/app-platform/ContentAppPlatform.cpp:503]: This part does not make sense to me.  We take a list.  Then for each entry in the list we create a length-1 list and write that to the binding attribute.  The outcome will be a binding list that just has the last entry from our initial list, no?  I doubt that's the goal here.

Was this code tested at all?

I suspect making `bindings` an `std::vector` and then creating a BindingListTypeInfo here from its `data()` and `size()` would work much better.

File: src/app/app-platform/ContentAppPlatform.h [hunk 8]
```diff
@@ -33,6 +33,8 @@ using chip::app::Clusters::ApplicationBasic::CatalogVendorApp;
 using chip::Controller::CommandResponseFailureCallback;
 using chip::Controller::CommandResponseSuccessCallback;
 
+using BindingListTypeInfo = chip::app::Clusters::Binding::Attributes::Binding::TypeInfo::Type;
```
Comment (bzbarsky-apple) [inline src/app/app-platform/ContentAppPlatform.h:36]:
```suggestion
using BindingListType = chip::app::Clusters::Binding::Attributes::Binding::TypeInfo::Type;
```
This is not a typeinfo in any meaningful sense.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/tv-app/linux/main.cpp (hunk 1 above)
Comment (chrisdecenzo) [reply examples/tv-app/linux/main.cpp]: that was WIP, and now removed

File: examples/tv-app/linux/main.cpp (hunk 2 above)
Comment (chrisdecenzo) [reply examples/tv-app/linux/main.cpp]: that was WIP, not removed

File: src/transport/raw/BLE.h (hunk 3 above)
Comment (chrisdecenzo) [reply src/transport/raw/BLE.h]: Addressed in https://github.com/project-chip/connectedhomeip/pull/15525

added comments explained rationale and behavior options

File: src/transport/raw/BLE.h (hunk 4 above)
Comment (chrisdecenzo) [reply src/transport/raw/BLE.h]: When there is more than one TransportManager, BLE only works for one of them. This change is meant to address that. Added comments in https://github.com/project-chip/connectedhomeip/pull/15525

File: src/transport/raw/BLE.h (hunk 6 above)
Comment (chrisdecenzo) [reply src/transport/raw/BLE.h]: Fixed naming in https://github.com/project-chip/connectedhomeip/pull/15525 and added comments explaining rationale and behavior.

**Comment by chrisdecenzo** [comment]
> Please split into 2 PRs: one for ACL and one for BLE.
> 
> ACL is reasonably clear - add a working ACL and some todos. BLE is unclear to me: what is the problem, how is this fix fixing it, how is a BLE transport supposed to work if it does not override receive? is it for send only (and if so maybe the methods should be 'send-only' or be some flag of send or send/receive or R/RW) or something else?

Split BLE changes out to https://github.com/project-chip/connectedhomeip/pull/15525

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 9]
```diff
@@ -388,14 +389,130 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+        // 0x050e = Account Login
+        std::list<ClusterId> allowedClusterList = { 0x0006, 0x0503, 0x0506, 0x0508, 0x0509, 0x050a, 0x050b };
+
+        for (const auto & clusterId : allowedClusterList)
+        {
+            Access::AccessControl::Entry::Target target = { .flags = Access::AccessControl::Entry::Target::kCluster |
+                                                                Access::AccessControl::Entry::Target::kEndpoint,
+                                                            .cluster  = clusterId,
+                                                            .endpoint = kLocalVideoPlayerEndpointId };
+            entry.AddTarget(nullptr, target);
```
Comment (mlepage-google) [inline src/app/app-platform/ContentAppPlatform.cpp:434]: You'll find that devices may support as few as 3 targets per entry. Example implementation does not have dynamic memory management, so without changing global config flags, or providing a different implementation, this will fail. You can however put several targets into one entry. There's also a limit on the number of entries per fabric which is not currently enforced (but will be), so you can't just make 3 entries for this (and still have entries for admin and other uses). Probably using a device type target is the answer here, but that is not yet supported.

File: examples/tv-app/android/include/application-basic/ApplicationBasicManager.h [hunk 10]
```diff
@@ -32,4 +32,6 @@ class ApplicationBasicManager : public ApplicationBasicDelegate
     uint16_t HandleGetProductId() override;
     CHIP_ERROR HandleGetApplicationVersion(AttributeValueEncoder & aEncoder) override;
     CHIP_ERROR HandleGetAllowedVendorList(AttributeValueEncoder & aEncoder) override;
+
+    std::list<uint16_t> GetAllowedVendorList() override { return { HandleGetVendorId(), 456 }; };
```
Comment (andy31415) [inline examples/tv-app/android/include/application-basic/ApplicationBasicManager.h:36]: what is special about 456? Please add a constant where name should be inituitive (e.g. kTestVendorId or similar)

File: examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h [hunk 11]
```diff
@@ -42,6 +42,8 @@ class ApplicationBasicManager : public ApplicationBasicDelegate
         mVendorId = vendorId;
         CopyString(mApplicationVersion, sizeof(mApplicationVersion), szApplicationVersion);
         mProductId = productId;
+        mAllowedVendorList.push_back(vendorId);
+        mAllowedVendorList.push_back(456); // test
```
Comment (andy31415) [inline examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h:46]: reuse the constant here.

File: examples/tv-casting-app/linux/main.cpp [hunk 12]
```diff
@@ -114,6 +115,30 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
 
 OptionSet * allOptions[] = { &cmdLineOptions, &helpOptions, nullptr };
 
+static void MyBoundDeviceChangedHandler(const EmberBindingTableEntry & binding, chip::DeviceProxy * peer_device, void * context)
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:118]: `My` prefix is a bit vague. Handlers are often named `On...`, so this could be `OnBoundDeviceChanged`

File: examples/tv-casting-app/linux/main.cpp [hunk 13]
```diff
@@ -114,6 +115,30 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
...
+{
+    using namespace chip;
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:120]: using namespace chip is already at the top level. Do we need these using statements?

File: examples/tv-casting-app/linux/main.cpp [hunk 14]
```diff
@@ -133,6 +158,9 @@ void PrepareForCommissioning(const Dnssd::DiscoveredNodeData * selectedCommissio
     // Display onboarding payload
     chip::DeviceLayer::ConfigurationMgr().LogDeviceConfig();
 
+    // Initialize binding handlers
+    InitBindingHandlers();
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:162]: Are these changes related to ACL?

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 15]
```diff
@@ -388,14 +389,130 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+    entry.SetAuthMode(Access::AuthMode::kCase);
+    entry.SetFabricIndex(kLocalFabricIndex);
+    entry.SetPrivilege(Access::Privilege::kOperate);
+    entry.AddSubject(nullptr, targetDeviceProxy->GetDeviceId());
+
+    std::list<Binding::Structs::TargetStruct::Type> bindings;
+
+    ChipLogProgress(Controller, "Create video player endpoint ACL and binding");
+    {
+        // 0x0006 = OnOff
```
Comment (andy31415) [inline src/app/app-platform/ContentAppPlatform.cpp:414]: add constants instead of comments

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 7 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp:502]: Sorry, this is a draft PR. Added that to the title.

Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: Updated to perform a single attribute write with the entire list

File: examples/platform/linux/AppMain.cpp [hunk 16]
```diff
@@ -588,7 +588,9 @@ void ChipLinuxAppMainLoop()
     SetDeviceAttestationCredentialsProvider(Examples::GetExampleDACProvider());
 
 #if CHIP_DEVICE_CONFIG_ENABLE_BOTH_COMMISSIONER_AND_COMMISSIONEE
+    ChipLogProgress(AppServer, "Starting commissioner\r\n");
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply examples/platform/linux/AppMain.cpp]: this is useful for debugging setup failures from logs, removed newlines

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 17]
```diff
@@ -388,14 +389,130 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+// the fabric index can't be 0, but needs to be validated as 1
+constexpr FabricIndex kLocalFabricIndex   = 1;
+constexpr ClusterId kNoClusterIdSpecified = kInvalidClusterId;
+
+CHIP_ERROR ContentAppPlatform::ManageClientAccess(OperationalDeviceProxy * targetDeviceProxy, uint16_t targetVendorId,
+                                                  NodeId localNodeId, Controller::WriteResponseSuccessCallback successCb,
+                                                  Controller::WriteResponseFailureCallback failureCb)
+{
+    Access::AccessControl::Entry entry;
+    GetAccessControl().PrepareEntry(entry);
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: added checks

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 9 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: updated configuration for TVs to have more entries and more targets

File: examples/tv-casting-app/linux/main.cpp (hunk 12 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: fixed name

File: examples/tv-casting-app/linux/main.cpp (hunk 13 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: removed

File: examples/tv-casting-app/linux/main.cpp (hunk 14 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp:163]: these relate to bindings. Updated title of PR

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 15 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: added constants

File: examples/tv-app/android/include/application-basic/ApplicationBasicManager.h (hunk 10 above)
Comment (bzbarsky-apple) [reply examples/tv-app/android/include/application-basic/ApplicationBasicManager.h:36]: If this is meant to be a test vendor id, it should use one of our spec-defined test values....

File: examples/tv-casting-app/linux/main.cpp [hunk 18]
```diff
@@ -114,6 +115,31 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
 
 OptionSet * allOptions[] = { &cmdLineOptions, &helpOptions, nullptr };
 
+static void OnBindingChanged(const EmberBindingTableEntry & binding)
+{
+    if (binding.type == EMBER_MULTICAST_BINDING)
+    {
+        ChipLogError(NotSpecified,
+                     "Group binding received nodeId=0x" ChipLogFormatX64 " remote endpoint=%d cluster=" ChipLogFormatMEI,
+                     ChipLogValueX64(binding.nodeId), binding.remote, ChipLogValueMEI(binding.clusterId.ValueOr(0)));
```
Comment (bzbarsky-apple) [inline examples/tv-casting-app/linux/main.cpp:124]: binding.nodeId is not valid for a multicast binding.  Neither is binding.remote.  This is a binding to a _group_, not a (node, endpoint) pair.

File: src/app/clusters/bindings/BindingManager.h [hunk 19]
```diff
@@ -93,6 +96,11 @@ class BindingManager
      */
     CHIP_ERROR NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster, void * context);
 
+    /*
+     * Notify the BoundDeviceChangedHandler that a binding has changed.
```
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.h:100]: Changed in what way?  This is not notifying removal of bindings, for example...

In general, what is the use case for this notification?  Right now it's not being used for anything, which makes it hard to see whether it makes sense at all.

Probably better to remove this notification for now, until it's actually being used.

File: examples/tv-app/tv-common/include/CHIPProjectAppConfig.h [hunk 20]
```diff
@@ -51,5 +51,10 @@
 // overrides CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT in CHIPProjectConfig
 #define CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT 16
 
+// For casting, we need to allow more ACL entries, and more complex entries
```
Comment (mlepage-google) [inline examples/tv-app/tv-common/include/CHIPProjectAppConfig.h:54]: OK but note this will expand storage usage, and RAM usage. Space is reserved for every entry, and all those targets and subjects in each one. You may wish to lower the subjects/targets limit to something like 8-10, keep the entries limit sufficiently high, and use more than one entry. There's no right answer here, it's mostly a question of balance. It's implemented this way for the smallest devices, ideally if we have more powerful devices, we can provide a dynamic implementation someday.

File: src/app/app-platform/ContentAppPlatform.h (hunk 8 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.h]: fixed

File: examples/tv-app/android/include/application-basic/ApplicationBasicManager.h (hunk 10 above)
Comment (chrisdecenzo) [reply examples/tv-app/android/include/application-basic/ApplicationBasicManager.h:36]: This was returned by the code before this change, and I was worried I would break the CI tests. Let me try removing it...

File: examples/tv-app/tv-common/include/CHIPProjectAppConfig.h (hunk 20 above)
Comment (chrisdecenzo) [reply examples/tv-app/tv-common/include/CHIPProjectAppConfig.h:54]: This is for a tv where long term we should have a dynamic implementation. But for now I just wanted to make sure we don't run out of space during testing.

File: src/app/clusters/bindings/BindingManager.h (hunk 19 above)
Comment (chrisdecenzo) [reply src/app/clusters/bindings/BindingManager.h]: Updated name to reflect usage. Added TODO in tv-casting-app indicating how this will be used in the next PR.

File: examples/tv-casting-app/linux/main.cpp (hunk 18 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: good catch, fixed

File: examples/tv-casting-app/linux/main.cpp [hunk 21]
```diff
@@ -114,6 +115,30 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
 
 OptionSet * allOptions[] = { &cmdLineOptions, &helpOptions, nullptr };
 
+static void OnBindingAdded(const EmberBindingTableEntry & binding)
+{
+    if (binding.type == EMBER_MULTICAST_BINDING)
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:120]: instead of == should we check for `!= EMBER_UNICAST_BINDING` instead?

If you do that, then:
  - we handle other types (e.g. EMBER_MANY_TO_ONE_BINDING ... or even UNUSED_BINDING if ever there)
  - can decrease the indent below
  
 Or do you plan to implement group binding support?
Why is this not a switch/ifelse to at least log_progress if an unknown/unsupported is received?

File: examples/tv-casting-app/linux/main.cpp [hunk 22]
```diff
@@ -114,6 +115,30 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
...
+    {
+        ChipLogError(NotSpecified, "Group binding received");
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:122]: Why is this LogError? The text does not specify. Text should contain "not yet implemented" or "not supported" and if not supported, a comment stating why (or if not yet implemented, a comment providing details in that case as well)

File: examples/tv-casting-app/linux/main.cpp [hunk 23]
```diff
@@ -114,6 +115,30 @@ HelpOptions helpOptions("tv-casting-app", "Usage: tv-casting-app [options]", "1.
...
+        return;
+    }
+
+    if (binding.type == EMBER_UNICAST_BINDING)
+    {
+        ChipLogError(NotSpecified,
+                     "Unicast binding received nodeId=0x" ChipLogFormatX64 " remote endpoint=%d cluster=" ChipLogFormatMEI,
```
Comment (andy31415) [inline examples/tv-casting-app/linux/main.cpp:129]: Add not implemented data in the error, since just seeing "ERROR: Unicast binding received nodeId=0x... remote endpoint=1 cluster=...." out of context is hard to figure out: why is it an error to receive a unicast binding? 

Something like "Unicast binding handling not implemented (received nodeid.....)"

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 24]
```diff
@@ -388,14 +389,139 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+// constexpr ClusterId kClusterIdAccountLogin        = 0x050e;
+
+CHIP_ERROR ContentAppPlatform::ManageClientAccess(OperationalDeviceProxy * targetDeviceProxy, uint16_t targetVendorId,
+                                                  NodeId localNodeId, Controller::WriteResponseSuccessCallback successCb,
+                                                  Controller::WriteResponseFailureCallback failureCb)
+{
+    Access::AccessControl::Entry entry;
+    ReturnErrorOnFailure(GetAccessControl().PrepareEntry(entry));
+    entry.SetAuthMode(Access::AuthMode::kCase);
+    entry.SetFabricIndex(kLocalFabricIndex);
```
Comment (andy31415) [inline src/app/app-platform/ContentAppPlatform.cpp:418]: should fabric index be passed in? How do we know 1 is always correct? The comments do not seem to help me: "the fabric index can't be 0, but needs to be validated as 1" just tells me that it cannot be 0 and I am not sure what "needs to be validated as 1" means.

File: src/app/clusters/bindings/BindingManager.cpp [hunk 25]
```diff
@@ -161,6 +161,15 @@ void BindingManager::FabricRemoved(CompressedFabricId compressedFabricId, Fabric
     mAppServer->GetCASESessionManager()->ReleaseSessionForFabric(compressedFabricId);
 }
 
+CHIP_ERROR BindingManager::NotifyBindingAdded(const EmberBindingTableEntry & binding)
+{
+    if (mBindingAddedHandler)
+    {
+        mBindingAddedHandler(binding);
```
Comment (andy31415) [inline src/app/clusters/bindings/BindingManager.cpp:168]: should the handler have the option to return a CHIP_ERROR? or do we need that notify always returns CHIP_NO_ERROR?

File: examples/tv-app/tv-common/include/CHIPProjectAppConfig.h (hunk 20 above)
Comment (mlepage-google) [reply examples/tv-app/tv-common/include/CHIPProjectAppConfig.h:54]: Understood.

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 26]
```diff
@@ -388,14 +389,139 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+// constexpr ClusterId kClusterIdApplicationLauncher = 0x050c;
+// constexpr ClusterId kClusterIdAccountLogin        = 0x050e;
+
+CHIP_ERROR ContentAppPlatform::ManageClientAccess(OperationalDeviceProxy * targetDeviceProxy, uint16_t targetVendorId,
+                                                  NodeId localNodeId, Controller::WriteResponseSuccessCallback successCb,
+                                                  Controller::WriteResponseFailureCallback failureCb)
+{
+    Access::AccessControl::Entry entry;
+    ReturnErrorOnFailure(GetAccessControl().PrepareEntry(entry));
+    entry.SetAuthMode(Access::AuthMode::kCase);
```
Comment (mlepage-google) [inline src/app/app-platform/ContentAppPlatform.cpp:417]: You should check error on all these calls:

- entry.SetAuthMode
- entry.SetPrivilege
- entry.AddSubject

It happens that short of using incorrect values, the first two will always succeed, with the current implementation (but maybe not with another?). The last one should succeed, but not if there isn't space (or out of memory etc.). Just always check the return values with one of those macros.

File: src/app/clusters/bindings/BindingManager.h [hunk 27]
```diff
@@ -93,6 +96,11 @@ class BindingManager
      */
     CHIP_ERROR NotifyBoundClusterChanged(EndpointId endpoint, ClusterId cluster, void * context);
 
+    /*
+     * Notify the BindingAddedHandler that a binding was added.
+     */
+    CHIP_ERROR NotifyBindingAdded(const EmberBindingTableEntry & binding);
```
Comment (bzbarsky-apple) [inline src/app/clusters/bindings/BindingManager.h:102]: Why do we want to do this via the BindingManager instead of directly registering this callback with the binding cluster?

And why do we think a single callback recipient is enough?

It might be better to just have a "bindings modified" event dispatched via the event system and then consumers can act on that as desired...

File: examples/tv-app/android/include/application-basic/ApplicationBasicManager.h (hunk 10 above)
Comment (chrisdecenzo) [reply examples/tv-app/android/include/application-basic/ApplicationBasicManager.h]: can't remove it or the CI tests fail.

File: examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h (hunk 11 above)
Comment (chrisdecenzo) [reply examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h]: made this a constant

File: src/app/clusters/bindings/BindingManager.cpp (hunk 25 above)
Comment (chrisdecenzo) [reply src/app/clusters/bindings/BindingManager.cpp:168]: there is no need for the callback to return a value since its a notification

File: src/app/clusters/bindings/BindingManager.h (hunk 27 above)
Comment (chrisdecenzo) [reply src/app/clusters/bindings/BindingManager.h:102]: It's helpful for the handler to receive the EmberBindingTableEntry object. What do you mean by "registering this callback with the binding cluster" - are you referring to the BindingTableAccess class?

can you point me to an example in the code where an event dispatched via the event system?

**Review by lazarkov — APPROVED** [review]

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 26 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: fixed

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 28]
```diff
@@ -418,9 +544,20 @@ CHIP_ERROR ContentAppPlatform::CreateBindingWithCallback(OperationalDeviceProxy
...
+
+    return CreateBindingWithCallback(device, deviceEndpointId, bindingList, successCb, failureCb);
+}
+
+CHIP_ERROR ContentAppPlatform::CreateBindingWithCallback(OperationalDeviceProxy * device, EndpointId deviceEndpointId,
+                                                         BindingListType bindingList,
+                                                         Controller::WriteResponseSuccessCallback successCb,
+                                                         Controller::WriteResponseFailureCallback failureCb)
+{
+    chip::Controller::BindingCluster cluster;
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: fixed

File: src/app/app-platform/ContentAppPlatform.cpp (hunk 24 above)
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp]: fixed to get actual fabric index

File: src/app/app-platform/ContentAppPlatform.cpp [hunk 29]
```diff
@@ -388,14 +389,139 @@ uint32_t ContentAppPlatform::GetPincodeFromContentApp(uint16_t vendorId, uint16_
...
+constexpr ClusterId kClusterIdContentLauncher = 0x050a;
+constexpr ClusterId kClusterIdAudioOutput     = 0x050b;
+// constexpr ClusterId kClusterIdApplicationLauncher = 0x050c;
+// constexpr ClusterId kClusterIdAccountLogin        = 0x050e;
+
+CHIP_ERROR ContentAppPlatform::ManageClientAccess(OperationalDeviceProxy * targetDeviceProxy, uint16_t targetVendorId,
+                                                  NodeId localNodeId, Controller::WriteResponseSuccessCallback successCb,
+                                                  Controller::WriteResponseFailureCallback failureCb)
+{
+    Access::AccessControl::Entry entry;
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply src/app/app-platform/ContentAppPlatform.cpp:417]: fixed

File: examples/tv-casting-app/linux/main.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: fixed

File: examples/tv-casting-app/linux/main.cpp (hunk 23 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: cleaned up

File: examples/tv-casting-app/linux/main.cpp (hunk 22 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: fixed (should have been progress and not error)

File: examples/tv-casting-app/linux/main.cpp (hunk 21 above)
Comment (chrisdecenzo) [reply examples/tv-casting-app/linux/main.cpp]: cleaned up

File: examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h [hunk 30]
```diff
@@ -41,7 +41,10 @@ class ApplicationBasicManager : public ApplicationBasicDelegate
         CopyString(mVendorName, sizeof(mVendorName), szVendorName);
         mVendorId = vendorId;
         CopyString(mApplicationVersion, sizeof(mApplicationVersion), szApplicationVersion);
-        mProductId = productId;
+        mProductId                    = productId;
+        static uint16_t kTestVendorId = 456;
```
[the remark this answers is not in the document]
Comment (chrisdecenzo) [reply examples/tv-app/linux/include/application-basic/ApplicationBasicManager.h]: fixed to use test vendor id

File: src/app/clusters/bindings/BindingManager.h (hunk 27 above)
Comment (chrisdecenzo) [reply src/app/clusters/bindings/BindingManager.h:110]: added code to return an error if more than one client attempts to register a callback. I can make this more flexible if the need arises.

Comment (bzbarsky-apple) [reply src/app/clusters/bindings/BindingManager.h:110]: > can you point me to an example in the code where an event dispatched via the event system?

`DeviceControlServer::CommissioningComplete` has an example.  But we can do that as a followup....

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/bindings/BindingManager.h (hunk 27 above)
Comment (bzbarsky-apple) [reply src/app/clusters/bindings/BindingManager.h:110]: > are you referring to the BindingTableAccess class?

I was referring to having the binding table expose some API to register change listeners, yes.  BindingManager is not that; it's the thing that manages actual CASE connections to bindings....
